openupgrade_scripts/scripts/loyalty/16.0.1.0/pre-migration.py

```diff
@@ -0,0 +1,806 @@
+            1,
+            1,
+            gc.create_uid,
+            gc.write_uid,
+            '[]',
+            gc.create_date,
+            gc.write_date
+        FROM gift_card AS gc
+        JOIN loyalty_program AS lp ON gc.program_id = lp.id
+        JOIN sale_order_line AS sol ON sol.gift_card_id = gc.id
+        """,
+    )
+    # Create records in loyalty_rule based on gift_card data
+    openupgrade.logged_query(
+        env.cr,
+        """
+        INSERT INTO loyalty_rrule (
+            program_id,
+            company_id,
+            minimun_qty,
+            create_uid,
+            write_uid,
+            product_domain,
+            reward_point_mode,
+            minimun_amount_tax_mode,
+            mode,
+            active,
+            reward_point_split,
+            create_date,
+            write_date,
+            reward_point_amount
+        )
+        SELECT
+            gc.program_id,
+            gc.company_id,
+            1,
+            gc.create_uid,
+            gc.write_uid,
+            '[]',
+            'money',
+            'incl',
+            'auto',
+            lp.active,
+            true,
+            gc.create_date,
+            gc.write_date,
+            1
+        FROM gift_card AS gc
+        JOIN loyalty_program AS lp ON gc.program_id = lp.id
+        """,
+    )
+    # If the card has been used, it calculates the new points value by subtracting the
+    # sum of the absolute price_unit values of the order lines linked to a gift_card
+    # from the initial value of the loyalty card (loyalty_card.initial_amount). In case
+    # there is no order line linked to the gift card, the initial value of the loyalty
+    # card shall be maintained.
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE loyalty_card lc
+        SET points =
+            CASE
+                WHEN (
+                    SELECT SUM(ABS(sol.price_unit))
+                    FROM sale_order_line sol
+                    JOIN gift_card gc ON gc.id = sol.gift_card_id
+                    WHERE gc.program_id = lc.program_id
+                ) IS NOT NULL
+                THEN lc.initial_amount - (
+                    SELECT SUM(ABS(sol.price_unit))
+                    FROM sale_order_line sol
+                    JOIN gift_card gc ON gc.id = sol.gift_card_id
+                    WHERE gc.program_id = lc.program_id
+                )
+                ELSE lc.points
+            END
+        FROM gift_card gc
+        WHERE lc.program_id = gc.program_id
+        """,
+    )
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    openupgrade.rename_fields(env, _field_renames)
+    openupgrade.rename_models(env.cr, _models_renames)
+    openupgrade.rename_tables(env.cr, _tables_renames)
+    openupgrade.rename_xmlids(env.cr, _xmlids_renames)
+    # Renamed in rename_xmlids method.
+    # In v15 are noupdate=1 and in v16 are noupdate=0
+    openupgrade.set_xml_ids_noupdate_value(
```

Comment:
The error message when testing the upgrade script is as follows:
```
odoo.exceptions.UserError: XML IDs mail_template_loyalty_card must be a tuple or list!
```

These lines of code can be changed to:
```python
    openupgrade.set_xml_ids_noupdate_value(
        env, "loyalty", ["mail_template_loyalty_card"], False
    )
    openupgrade.set_xml_ids_noupdate_value(
        env, "loyalty", ["gift_card_product_50"], False
    )
```

Reply:
thanks